lib/request.js

```diff
@@ -1,19 +1,21 @@
 'use strict'
 
-function Request (params, req, body, query, log) {
+function Request (params, req, body, query, headers, log) {
```

Comment:
Why pass in the headers as a parameter when you can simply `this.headers = req.headers`?

Reply:
Because `this.req.headers` can be undefined.  That's the result of testing. I didn't checked why it happens.

Reply:
I said `req.headers` not `this.req.headers`. If the original incoming request, which is what `req` is doesn't have any headers then it isn't a valid HTTP 1.1 request.

Reply:
Correct me If I'm wrong but with your approach I get lots of `Cannot read property ``headers` of undefined` messages because at this place https://github.com/StarpTech/fastify/blob/feature/%23200_%23191_headers_shortcut_schema_validation/fastify.js#L248 we build a request object without passing the parameters. I think this construct belongs to the performance optimization. Passing headers as an argument work fine.

Reply:
```
diff --git a/lib/handleRequest.js b/lib/handleRequest.js
index b921e41..ba5bdb5 100644
--- a/lib/handleRequest.js
+++ b/lib/handleRequest.js
@@ -11,7 +11,7 @@ function handleRequest (req, res, params, store) {
   var method = req.method
 
   if (method === 'GET' || method === 'DELETE' || method === 'HEAD') {
-    return handler(store, params, req, res, null, urlUtil.parse(req.url, true).query, req.headers)
+    return handler(store, params, req, res, null, urlUtil.parse(req.url, true).query)
   }
 
   if (method === 'POST' || method === 'PUT' || method === 'PATCH') {
@@ -42,7 +42,7 @@ function handleRequest (req, res, params, store) {
       setImmediate(wrapReplyEnd, req, res, 415)
       return
     }
-    return handler(store, params, req, res, null, urlUtil.parse(req.url, true).query, req.headers)
+    return handler(store, params, req, res, null, urlUtil.parse(req.url, true).query)
   }
 
   setImmediate(wrapReplyEnd, req, res, 405)
@@ -77,11 +77,11 @@ function jsonBodyParsed (err, body, req, res, params, handle) {
     setImmediate(wrapReplyEnd, req, res, 422)
     return
   }
-  handler(handle, params, req, res, body, urlUtil.parse(req.url, true).query, req.headers)
+  handler(handle, params, req, res, body, urlUtil.parse(req.url, true).query)
 }
 
-function handler (store, params, req, res, body, query, headers) {
-  var request = new store.Request(params, req, body, query, headers, req.log)
+function handler (store, params, req, res, body, query) {
+  var request = new store.Request(params, req, body, query, req.log)
   var valid = validateSchema(store, request)
   if (valid !== true) {
     setImmediate(wrapReplyEnd, req, res, 400, valid)
diff --git a/lib/request.js b/lib/request.js
index 6b30c97..4fb7def 100644
--- a/lib/request.js
+++ b/lib/request.js
@@ -1,6 +1,7 @@
 'use strict'
 
-function Request (params, req, body, query, headers, log) {
+function Request (params, req, body, query, log) {
+  const headers = (req && req.headers) ? req.headers : {}
   this.params = params
   this.req = req
   this.body = body
@@ -10,7 +11,8 @@ function Request (params, req, body, query, headers, log) {
 }
 
 function buildRequest (R) {
-  function _Request (params, req, body, query, headers, log) {
+  function _Request (params, req, body, query, log) {
+    const headers = (req && req.headers) ? req.headers : {}
     this.params = params
     this.req = req
     this.body = body
diff --git a/test/internals/handleRequest.test.js b/test/internals/handleRequest.test.js
index a9f0b68..dd2a7e4 100644
--- a/test/internals/handleRequest.test.js
+++ b/test/internals/handleRequest.test.js
@@ -23,13 +23,13 @@ function schemaCompiler (schema) {
 
 test('Request object', t => {
   t.plan(7)
-  const req = new Request('params', 'req', 'body', 'query', 'headers', 'log')
+  const req = new Request('params', 'req', 'body', 'query', 'log')
   t.type(req, Request)
   t.equal(req.params, 'params')
   t.deepEqual(req.req, 'req')
   t.equal(req.body, 'body')
   t.equal(req.query, 'query')
-  t.equal(req.headers, 'headers')
+  t.deepEqual(req.headers, {})
   t.equal(req.log, 'log')
 })
 ```

Reply:
With an additional check, it works but I want to avoid the additional checks but in this way we need to pass less across fastify I will refactor it thanks.